Refactor ty::FnSig to contain a &'tcx Slice<Ty<'tcx>>

Mark-Simulacrum opened the pull request:
We refactor this in order to achieve the following wins:

 - Decrease the size of `FnSig` (`Vec` + `bool`: 32, `&Slice` + `bool`: 24).
 - Potentially decrease total allocated memory due to arena-allocating `FnSig` inputs/output; since they are allocated in the type list arena, other users of type lists can reuse the same allocation for an equivalent type list.
 - Remove the last part of the type system which needs drop glue (#37965 removed the other remaining part). This makes arenas containing `FnSig` faster to drop (since we don't need to drop a Vec for each one), and makes reusing them without clearing/dropping potentially possible.

r? @eddyb

eddyb commented on src/librustc/ty/relate.rs:
```diff
@@ -185,32 +187,28 @@ impl<'tcx> Relate<'tcx> for ty::FnSig<'tcx> {
-                      variadic: a.variadic})
+        let inputs_and_output = a.inputs().into_iter().cloned()
+            .zip(b.inputs().into_iter().cloned())
+            .zip(iter::repeat(false))
```
This can be done with `.map`.

eddyb commented on src/librustc/ty/structural_impls.rs:
```diff
@@ -589,17 +587,23 @@ impl<'tcx> TypeFoldable<'tcx> for ty::TypeAndMut<'tcx> {
+            .map(|i| i.fold_with(folder))
+            .collect::<AccumulateVec<[_; 8]>>();
+        ty::FnSig {
+            inputs_and_output: folder.tcx().intern_type_list(&inputs_and_output),
```
I think you can just fold the list wholesale.

eddyb commented on src/librustc/ty/sty.rs:
```diff
@@ -563,22 +563,31 @@ pub struct ClosureTy<'tcx> {
 impl<'tcx> PolyFnSig<'tcx> {
-    pub fn inputs(&self) -> ty::Binder<Vec<Ty<'tcx>>> {
-        self.map_bound_ref(|fn_sig| fn_sig.inputs.clone())
+    pub fn inputs<'a>(&'a self) -> Binder<&[Ty<'tcx>]> {
```
The elision abuse here looks odd.

eddyb commented on src/librustc/ty/sty.rs:
```diff
@@ -1243,10 +1252,6 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
 
     // Type accessors for substructures of types
-    pub fn fn_args(&self) -> ty::Binder<Vec<Ty<'tcx>>> {
-        self.fn_sig().inputs()
```
You could keep this returning a slice.

eddyb commented on src/librustc_trans/callee.rs:
```diff
@@ -491,13 +500,12 @@ fn trans_fn_pointer_shim<'a, 'tcx>(
-    };
+    let tuple_input_ty = tcx.intern_tup(sig.inputs());
+    let sig = tcx.mk_fn_sig(
+        iter::once(bare_fn_ty_maybe_ref).chain(iter::once(tuple_input_ty)),
```
You're using `iter::once` a bit too much - try arrays instead.

eddyb commented on src/librustc_trans/common.rs:
```diff
@@ -1091,10 +1091,11 @@ pub fn ty_fn_ty<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
-                ..sig
-            });
+            let sig = sig.map_bound(|sig| tcx.mk_fn_sig(
+                iter::once(env_ty).chain(sig.inputs().into_iter().cloned()),
```
Could you avoid `.into_iter()` on slices? It's just `.iter()`.

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -1595,7 +1595,8 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
                 let inputs = bare_fn_ty.sig.inputs();
-                let late_bound_in_args = tcx.collect_constrained_late_bound_regions(&inputs);
+                let late_bound_in_args = tcx.collect_constrained_late_bound_regions(
+                    &inputs.map_bound(|i| i.to_owned()));
```
Why would this be needed? Looks like a bug?

Mark-Simulacrum replied:
`TypeFoldable` isn't implemented for `&[Ty<'tcx>]`, which is what we have, so we need to cast it to a vector in order to pass it here. This is rather unfortunate, but I'm not sure how much we can do. It may be worth pointing out that the use is a visit_with on it, so it's possible this can be refactored into passing either the slice or an iterator to solve this problem. I can look into that if you'd like.

KalitaAlexey commented:
What is purpose of this?

Mark-Simulacrum commented:
I believe there are a couple of purposes:
 - Decreasing the size of `FnSig`
   - `Vec` + `bool`: 32, `&Slice` + `bool`: 24.
 - Potentially decrease total allocated memory due to arena-allocating `FnSig` inputs/output; since they are allocated in the type list arena, other users of type lists can reuse the same allocation for an equivalent type list.
 - Remove the last part of the type system which needs drop glue (#37965 removed the other remaining part)

@eddyb may be able to suggest other reasons.

---

@eddyb Ready for another review pass.

bors commented:
:umbrella: The latest upstream changes (presumably #38053) made this pull request unmergeable. Please resolve the merge conflicts.

Mark-Simulacrum commented:
Rebased.

eddyb commented on src/librustc_typeck/lib.rs:
```diff
@@ -274,14 +271,12 @@ fn check_start_fn_ty(ccx: &CrateCtxt,
-                }),
+                sig: ty::Binder(tcx.mk_fn_sig(
+                    iter::once(tcx.types.isize)
+                        .chain(iter::once(tcx.mk_imm_ptr(tcx.mk_imm_ptr(tcx.types.u8)))),
```
Could you use slice iterators, here and elsewhere?

eddyb commented:
@Mark-Simulacrum Can you move the contents of that comment into the description? You may also want to mention that the drop glue thing is about arenas, both in terms of cost of dropping/reusing the temporary ones currently and experiments with unified arenas in the future.

eddyb commented:
@rust-lang/compiler Any objections? This is ready to land otherwise, modulo a few "style" tweaks.

Mark-Simulacrum commented:
Fixed style nit (I couldn't find any other places I chained iter::once together).

Added the comment for reasons why we want this to the description.

bors commented:
:umbrella: The latest upstream changes (presumably #38121) made this pull request unmergeable. Please resolve the merge conflicts.

Mark-Simulacrum commented:
Rebased.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 296ec5f has been approved by `eddyb`

<!-- @bors r=eddyb 296ec5f9b7bf5aec05ed6672d3499079d35e4594 -->

bors commented:
:hourglass: Testing commit 296ec5f with merge ff261d3...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/11208), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/11271), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/2495), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/2314), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/4482), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/11213), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/11161), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/3140), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/11135), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/4363), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/6313), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/11254), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/11228), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/3173), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/2353), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/6356), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/3207), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/6319), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/5505), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2517), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/6331), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/3196)
Approved by: eddyb
Pushing ff261d3a6b5964e1e3744d055238de624afc5d76 to master...

michaelwoerister commented:
Nice!
